src/WebdriverClassicDriver.php

```diff
@@ -136,16 +136,22 @@ public function stop(): void
      */
     public function reset(): void
     {
+        $webDriver = $this->getWebDriver();
+
         // switch to default window..
-        $this->switchToWindow();
+        $initialWindow = substr((string)$this->initialWindowName, strlen(self::W3C_WINDOW_HANDLE_PREFIX));
```

Comment:
This will still not work because `initialWindowName` may contain a name and not a handle, since `$this->getWindowName()` can return a window name (instead of handle). The substr would end up slicing out a part of the name.

Reply:
Here's a patch, it fixes several things:
```patch
Subject: [PATCH] use window handles internally
---
Index: src/WebdriverClassicDriver.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/WebdriverClassicDriver.php b/src/WebdriverClassicDriver.php
--- a/src/WebdriverClassicDriver.php	(revision 71c242b36e92fe0fe3891b2700a41eb629b61550)
+++ b/src/WebdriverClassicDriver.php	(date 1712604857594)
@@ -78,7 +78,7 @@
 
     private string $webDriverHost;
 
-    private ?string $initialWindowName = null;
+    private ?string $initialWindowHandle = null;
 
     /**
      * @param string $browserName One of 'edge', 'firefox', 'chrome' or any one of {@see WebDriverBrowserType} constants.
@@ -104,7 +104,7 @@
         try {
             $this->webDriver = RemoteWebDriver::create($this->webDriverHost, $this->desiredCapabilities);
             $this->applyTimeouts();
-            $this->initialWindowName = $this->getWindowName();
+            $this->initialWindowHandle = $this->webDriver->getWindowHandle();
         } catch (\Throwable $e) {
             throw new DriverException("Could not start driver: {$e->getMessage()}", 0, $e);
         }
@@ -136,13 +136,17 @@
      */
     public function reset(): void
     {
-        // switch to default window..
-        $this->switchToWindow();
-        // ..and close all other windows
-        foreach ($this->getWindowNames() as $name) {
-            if ($name !== $this->initialWindowName) {
-                $this->withWindow($name, fn() => $this->getWebDriver()->close());
-            }
+        $webDriver = $this->getWebDriver();
+
+        // close all windows except the initial one
+        foreach ($webDriver->getWindowHandles() as $windowHandle) {
+            if ($windowHandle === $this->initialWindowHandle) {
+                continue;
+            }
+
+            $webDriver->switchTo()->window($windowHandle);
+            $webDriver->close();
+            $this->switchToWindow();
         }
 
         $this->getWebDriver()->manage()->deleteAllCookies();
@@ -175,13 +179,9 @@
 
     public function switchToWindow(?string $name = null): void
     {
-        if ($name === null) {
-            $name = $this->initialWindowName;
-        }
-
-        if (is_string($name)) {
-            $name = $this->getWindowHandleFromName($name);
-        }
+        $name = $name === null
+            ? $this->initialWindowHandle
+            : $this->getWindowHandleFromName($name);
 
         $this->getWebDriver()->switchTo()->window((string)$name);
     }
```

Reply:
> This will still not work because `initialWindowName` may contain a name and not a handle, since `$this->getWindowName()` can return a window name (instead of handle). The substr would end up slicing out a part of the name.

When the initial window name is populated (after the session was just started with an empty window and no URL visited yet) the window doesn't have a name. It even doesn't have any HTML in it. So I'm pretty certain that `substr` would work.

> Here's a patch, it fixes several things:

I like it. I've made several adjustments to:

* account for code changes since the patch creation time;
* keep initial `$this->switchToWindow();` call, because it would save the day when the test finishes at the point when the currently active window is removed and WebDriver would throw the `NoSuchWindow` exception on any command we perform.

Reply:
> keep initial $this->switchToWindow(); call, because it would save the day when the test finishes at the point when the currently active window is removed and WebDriver would throw the NoSuchWindow exception on any command we perform.

Aah, I see. If there's only one window, then the foreach loop would skip switching to it and might stay stuck on a closed window. Maybe this would be a cleaner solution (and perhaps also better performing)?
```php
foreach ($webDriver->getWindowHandles() as $windowHandle) {
    if ($windowHandle !== $this->initialWindowHandle) {
        continue;
    }
    $webDriver->switchTo()->window($windowHandle);
    $webDriver->close();
}
$this->switchToWindow();
```

Reply:
I've played around a bit and surprisingly:

* the `getWindowHandle` call made inside a closed window context fails with a `NoSuchWindow` exception;
* the `getWindowHandles` call made inside a closed window context succeeds;
* attempts to switch windows from inside a closed window context succeeds.

This proves, that calling `$this->switchToWindow();` before/inside the loop isn't necessary.

I've updated the PR code to include the proposed changes.